### ISSUE TITLE
Request: Display splice junction prevalence on gene page and/or JBrowse

### ISSUE THREAD

**Issue by chris-grove** [WormBase/website#6296]
At the 2017 International C. elegans Meeting, there was a presentation by Denis Dupuy in which his lab had developed software to calculate (using RNAseq data) and display the prevalence of various splice junctions for genes with different splice variants/isoforms. The point of this type of display is to present to users which splice variants are actually abundant or have a significant presence in vivo in contrast to splice variants that are only present in a small fraction of the reads. The paper from the Dupuy lab came out in December:

https://www.ncbi.nlm.nih.gov/pubmed/29089372

This set off an email thread between Denis, @gw3 @khowe @scottcain and myself.

There was discussion about other potential technologies to achieve the same goal, such as LeafCutter:

Paper:
https://www.ncbi.nlm.nih.gov/pubmed/29229983

Site:
https://davidaknowles.github.io/leafcutter/

I'm creating this ticket to keep track of the conversation and any ideas on implementation that we might come up with. I think this is an important feature, particularly for complex loci with 10+ isoforms:

http://www.wormbase.org/species/c_elegans/gene/WBGene00006759#09g--10
http://www.wormbase.org/species/c_elegans/gene/WBGene00006767#09g--10
http://www.wormbase.org/species/c_elegans/gene/WBGene00003150#09g--10

Of course, whatever technology we go with we would want it to be able to be run automatically during the build with updated RNA-seq data.

**Comment by chris-grove** [comment]
I'll just add that this was specifically requested at the Baltimore Worm Club where I recently gave a WormBase talk

**Comment by chris-grove** [comment]
Just discussed with @khowe 
Until we can arrive at a longer term, more sustainable solution, we think it would be good to create a section in the Expression widget that displays the static images generated by Denis' lab algorithm, possibly called "Splice junction prevalence" or something similar. Here's a link to the PDF document with all gene images (Supplemental Figure S2 from the paper):

https://www.dropbox.com/s/ipeuxzvjwn0yaxr/Supplemental_Fig_S2.pdf?dl=0

I'm adding the Webteam label. I'd be happy to discuss implementation details.

**Comment by epaule** [comment]
Couldn't we just use the score column in the RNASeq intron track and do a
colour gradient? Lot simpler, more compact and shows the same as having a
separate axis.

**Comment by scottcain** [comment]
I've never been a fan of that intron track either--I think it is pretty hard to interpret.  I suspect my suggested solution (inferring the frequency of resulting transcripts) is probably not possible (bummer, as it would be really informative and easy to create).  But a track that incorporated vertical sorting and color gradients might get the job done.

**Comment by gw3** [comment]
@chris-grove  I agree that the RNASeq Introns track is not nice and could be modified.

A long time ago when it was first being implemented, I suggested that it could be displayed with the thickness of the blocks being proportional to the log of the number of reads, but this appears to have been forgotten over time :-)

It would be good it the Introns could also be sorted so that the strong ones appear at the top.

**Comment by scottcain** [comment]
In JBrowse it's really hard (read: currently impossible) to sort unrelated features.  One possible workaround that could possibly work in this situation is if the features we want to compare all have the same parent (like a gene).  I don't think the RNAseq intron data are arranged that way, but I would think it would be possible to make that happen.  Then I could write a new glyph that subclasses the gene glyph and add sorting to it.

And yes, adding log of the score would help a lot.  I wasn't here when that track was written :-)  I'll look into adding that for 265.

**Comment by gw3** [comment]
In a recent Web Apollo gene curation project, we split up the RNASeq Introns into different tracks depending on their score, so the user could choose to just turn on high-scoring tracks, or high and medium-scoring tracks, etc.

**Comment by scottcain** [comment]
Hi @gw3 ,

Having a separate high scoring track is a good idea to; what was the cut off?  I could also deal with strand in any of the ways you suggested--which would you prefer?

**Comment by gw3** [comment]
How about, for elegans:
Low: 0 to 99
Medium: 100 to 9999
High: 10000+

And for other species:
Low: 0 to 9
Medium: 10 to 99
High 100+

Or maybe just (elegans) Low: 0-999 High: 1000+
(other species) Low: 0-49 High: 50+

We can play around with the thresholds if these are not informative

Having an arrow-head might make it hard to determine the exact splice site.
How about have forward and reverse introns in the same pink and cyan that the curated gene structures are displayed in?

**Comment by scottcain** [comment]
Back to the topic that this issue is about: Denis Dupuy's data.  I got some sample data for several thousand genes (which may in fact be the complete data set) from him and created this for three actin genes:

https://staging.wormbase.org/tools/genome/jbrowse-simple/?data=data%2Fc_elegans_PRJNA13758&loc=V%3A11070151..11083350&tracks=Curated_Genes%2Cdemo%20splice%20juntions&highlight=

These data are a little different than I expected: there are few introns annotated per gene and so the display would not be anywhere near as complicated as I expected.  The scores are raw read counts, which for the data provided by Denis ranges from 1M to 15M reads.  I created a simple gradient that covers that range with light blue on the low end to a dark blue on the high end.  Assuming this is representative of the entire data set, I don't really think there needs to be much more, though obviously I'm open to suggestions.  The track still needs better labeling and a good "about" statement with a link back to Denis' website (which I assume exists :-)

**Comment by chris-grove** [comment]
@scottcain 
That's a possibility. One thing I really like about Denis' plots are that you can quickly see which splice junctions represent 1-100% of splice isoforms, as well as being able to see what can be considered "noise" (<1% prevalence) which can account for, or explain, why some transcript isoforms are visible. For example, the act-1 gene has four visible transcripts, T04C12.6.1 having extra exons, which in Denis' plots we can see fall below the 1% prevalence threshold, but which appear to be unfounded in the JBrowse track. Also, the vertical axis representing log-scale prevalence is a significant, and I think helpful, feature of the visualization.

Denis has a website:

http://www.iecb.u-bordeaux.fr/teams/DUPUY/DupuylabSite/Welcome.html

and he has a link to the large supplemental figure from the paper:

http://www.iecb.u-bordeaux.fr/teams/DUPUY/DupuylabSite/Publications.html

If that is appropriate.

**Comment by scottcain** [comment]
@gw3 ,
I created four new RNAseq intron tracks: one that uses log scaling and uses the forward/reverse gene coloring but is otherwise like the existing intron track, and three tracks for high, medium and low scoring introns.  Take a look at:

https://staging.wormbase.org/tools/genome/jbrowse-simple/full.html?data=data%2Fc_elegans_PRJNA13758&loc=X%3A12963341..12972081&tracks=Curated_Genes%2Chigh%20scoring%20RNAseq%20introns%2Cmedium%20scoring%20RNAseq%20introns%2Clow%20scoring%20RNAseq%20introns%2Ctest%20RNASeq%20introns&highlight=

and let me know what you think.

**Comment by scottcain** [comment]
Things that I might think about doing for these RNAseq intron tracks:

* Using variations of color to indicate the high, medium and low tracks visually
* Removing the score from at least the high scoring track, and maybe all of them, replacing with a mouseover that indicates the read count (removing the score label would make the features more compact and thus more likely to fit on a single line)

**Comment by Paul-Davis** [comment]
Hi, 

I like these new tracks.

I was just looking at the new tracks with @gw3 and I think my preferred option would be a combination of both approaches.

Split them into 2 Tracks "Hi Scoring RNASeq Introns" (1-999) and "Low Scoring RNASeq Introns" (1000+) and colour and score as the Test track that is up at the moment.

This gives clarity of display and allows for filtering of the poor quality data like we do in our curation databases.

Paul

**Comment by gw3** [comment]
@scottcain Many thanks for doing this.
- adding colour to indicate strand is a huge improvement. Keep this.
- having the three high, medium and low groups makes it easy to filter and sort the introns. It would be interesting to see what having just two groups feels like.
- the blocks still look too thick (high) and so use up too much valuable screen-space. Can they be scaled down to be thinner?
- I think it is worth trying your idea of shading the colour to indicate score.

**Comment by chris-grove** [comment]
@scottcain 
Yes, I like the three tracks (low, medium, high); agree it would be worth seeing two tracks (high, low). 

Also agree that the color/hue to indicate the strand is helpful, the glyphs could be thinner/shorter, and color/shade would be good to indicate score/#reads, possibly with mouseover caption to indicate # reads.

Would it be possible to normalize what is considered "high", "medium" and "low" for a given gene/locus? Right now, many gene models completely lack any representation in the "high" track and only have representation in the "medium" and "low" tracks.

**Comment by scottcain** [comment]
@chris-grove and @gw3  JBrowse can't really do on the fly normalization like that; it would require some precomputing.  I can imagine writing a script to do that analysis (read in gene gff, saving boundaries; read in introns, counting those that fall within boundaries; looping through each gene and calculating a normalized read count; spit out gff).  If such a thing existed, would it have to be redone with each release, or do these intron data get added to on regular (or irregular) basis?

**Comment by gw3** [comment]
Normalisation within a gene assumes that we know the boundaries of the gene.
I would prefer to not do this as we are sometimes not certain about the gene boundaries, especially in non-elegans species.
I regard the intron data as a valuable set of evidence to confirm the curated gene structure. Using the gene structure to determine how that evidence is displayed comes dangerously close to being a circular-reasoning fallacy.

**Comment by chris-grove** [comment]
@gw3 @scottcain 
OK, yes, normalization isn't the way to go. Could we possibly change the title of the track from "RNASeq Introns" to "Splice Junctions" or something like that? I wonder how discoverable the track is as it is currently named. I understand why it would be named like that (it makes sense) but I wonder if users pick up on that when browsing track names.

**Comment by gw3** [comment]
@chris-grove @scottcain
Yes, I like the idea of renaming the track to  "Splice Junctions".

**Comment by scottcain** [comment]
Hi @gw3 and @chris-grove I have another test track that is still in development, but I thought I'd show you how things stand.  I got a small set of data from Denis for the first 100kb or so of I.  I created a glyph that will take splice junctions that are attributed to a gene and display them together, sorting the high scoring ones to the top of the group and color them darker.  I have a few things I'd still like to do with this glyph, like better packing of the individual glyphs, which may or may not be possible (for example, putting them all on the same line would imply that they all scored the same, which they did not), and possibly adding labeling that would indicate the score (mouseovers won't work in this instance since each gene grouping is effectively a single glyph from a mouseover perspective).  I'm not sure that will work either since it will make an already talk track much taller.  Anyway, this is something to take pot shots at: https://staging.wormbase.org/tools/genome/jbrowse-simple/full.html?data=data%2Fc_elegans_PRJNA13758&loc=I%3A1967..19118&tracks=Curated_Genes%2Cdupuy_test

**Comment by scottcain** [comment]
I should add that the color cut off for this glyph is a simple >1000 and color it darker.  It would certainly be possible to do something more sophisticated--in this instance, normalizing might make sense since it's on a gene by gene basis.

**Comment by chris-grove** [comment]
Thanks @scottcain , I'll have a look and get back to you

**Comment by gw3** [comment]
HI @scottcain, I far prefer the RNASeq intron tracks to the Dupuy data, because it is not processed so much and so you can spot subtleties in the expression pattern that are often useful when curating gene, especially in unusual elegans gene structures and in non-elegans species where much of our focus is now placed. I find in practice that when doing curation, the Dupuy data takes up far too much screen space and it is easy to miss things in it.

**Comment by chris-grove** [comment]
@scottcain @gw3 
I think it should be emphasized that we may be discussing two (or more) different use cases: (1) gene model curation and (2) users wanting to visualize splice junction prevalence for a particular gene, especially for genes with many transcript isoforms (see C. elegans ant-1.1, for example). 

Some of the main reasons Denis' display is useful is:
1) It displays the splice junction prevalence in a fixed space, on a log10 scale vertically, quickly giving the visual indication as to what are substantially represented junctions, and which are likely noise (for example quickly visually discerning junctions that represent <1% of transcript reads (which most people probably want to filter out), 1%-10%, and 10%-100%)
2) Normalization to the gene; the first point above relies on normalization, which I understand isn't always perfect, but I think it provides a good deal of benefit to users. As long as we're not making calls about the gene model based on the normalization, I don't think we run the risk of circular reasoning.
3) The splice junction glyphs are placed visually close to the "core" gene model making it easy to see, in relation to the gene model, where the different junctions lie

Given the advantages I've just listed, and the challenges in creating a JBrowse track described earlier, I'm inclined to vote for displaying the static images on gene pages, if not instead of a JBrowse track, at least as a complementary view. I just have a hard time imagining a genome browser track that can provide all of these benefits, particularly the compact display with junctions visually close to the core gene model. Maybe there are other options that I haven't considered.

**Comment by scottcain** [comment]
I've updated the JBrowse instance on staging to have new splice junction tracks: one for less than 100 reads and one for more than 100.  In the "more than 100" track (which I've labeled "RNASeq Splice Junctions (common)" also has glyphs get darker as there are more reads.  See

https://staging.wormbase.org/tools/genome/jbrowse-simple/full.html?data=data%2Fc_elegans_PRJNA13758&loc=II%3A3294021..3312120&tracks=Curated_Genes%2CRNASeq%20Splice%20Junctions%20(common)%2CRNASeq%20Splice%20Junctions%20(rare)

for example.

As for the images created by Denis', we would have to decide how to present them in the context of a gene page.  I don't have much of an opinion.  Presumably, we would pull them in dynamically from where ever Denis' group is hosting them.  Should they be in their own widget perhaps?

**Comment by gw3** [comment]
Hi @scottcain, this looks very good!

I would have the threshold a bit higher, we use a threshold of 1000 for gene curation, but I think maybe a threshold of 500 would work better in this browser.

Could the dynamic range of the colour scheme be increased? At present there is just a slight difference in shade between introns with 10^2 reads and introns with 10^6 reads.

But even without these changes, it it good for use on elegans.

I tried looking at some other species, but nothing was displayed. Have you loaded the data for other species yet? You will need a different threshold for each species.

**Comment by chris-grove** [comment]
@scottcain  @gw3 
Thanks for that. I think one thing I find confusing (or think users would find confusing) is that there are so many splice junctions that suggest the current gene models (displayed in the Curated Genes track) ought to be updated to accommodate the "common" splice junctions. See for example the locus you linked to in your JBrowse link:

![screen shot 2018-06-11 at 10 30 03 am](https://user-images.githubusercontent.com/611382/41237884-b02d9a8c-6d62-11e8-9af1-0cebb3f9da9a.png)

I see one splice junction that apparently is represented by 27,000+ reads but isn't reflected in the gene models of the Curated Genes track.

**Comment by scottcain** [comment]
@chris-grove I wondered about that too, but figured there was likely a good explanation :-)

@gw3 I agree with your comments about the dynamic range for the colors--I was reusing a function that I used elsewhere that worked well but this needs tweaked more.  Do you have the cut offs that you use in curation for the other species?

**Comment by gw3** [comment]
@chris-grove it is a common artefact of short-read alignments that they will align across widely separated regions where there are duplicated gene so that they match their first half to one duplicate and the second half to to second duplicate. (This is a prime heads-up indicator that I use to notice duplicated genes when curating.) This accounts for most of the strong introns that are not used when curating gene structures.

It your indicated region, there is something else going on. :-)

There has been an insertion of a MARINER3 transposon (F39E9.5) that appears to have resulted in the disruption of the gene or genes that used to be in this region. 
The resulting wreckage has been curated as three pseudogenes.
We don't put huge amounts of effort into reconstructing a plausible ancestral gene structure when the damage is as severe as it is in this region. It looks like three gene regions, so we have left it as that, but there is a region that can be linked to from F39E9.28 using your 27k intron.
Now that I have looked at this region for a while, I can make out a hint of the structure of the parent coding gene (F55A4.5) that this region is a duplicate of. The reads expressed from the coding parent gene are being aligned to this pseudogenic region and the 27K intron is a part of the structure of the duplicate gene that still remains.
This region could be curated to reconstruct the original gene's structure, but it would take 30 minutes or so.

There may be something interesting going on in this region as there is a cluster of tncRNA genes that appear to be processed from  the exons of the pseudogene F39E9.7.
The coding parent of the pseudogene F39E9.7 is F55A4.5, and there are no tncRNA genes annotated in the corresponding exon of F55A4.5.
It is VERY rare (we know of one published example) for there to be any functional phenotypic activity associated with a pseudogene, so I would be surprised if the annotated tncRNA genes were functional.
I rather suspect that there is no expression occurring in your region and that all expression is really going on in the coding parent gene F55A4.5 and that the tncRNA gene cluster has been annotated in the wrong location and should be placed in the seventh exon of the coding parent gene F55A4.5.

@Paul-Davis does this sound a reasonable interpretation?

**Comment by Paul-Davis** [comment]
It is likely that the tnRNAs are false positives and the miRNA that also aligns there was also questioned and removed from miRBase (something I have on my list to do).

**Comment by gw3** [comment]
@scottcain let's try a cut-off of 20 for all non-elegans species. We can always adjust this in the future if people object.

**Comment by chris-grove** [comment]
@scottcain @gw3 
OK, I can see that these are artefacts and are useful considerations for curating gene structures, but I think these things would only be a source of confusion for users. It seems to me that it would be good for users to have access to the prevalence of splice junctions that are actually included in the proposed gene models of the Curated Genes track, otherwise I think it will turn off most users.

**Comment by scottcain** [comment]
Re: Color variation.  I was using the log of the read count to vary the color which obviously flattens the differences quite a bit.  I changed it to vary linearly up to 20,000 reads, when the color is fully saturated and pretty dark (though I could be convinced to go a little darker).

I also added the other species that have these data.  I'm using the same function where the max is 20,000, but that means that most glyphs are pretty light in color in those species.  I'll have to create separate configs for non-elegans species which will have to include different color functions as well as different descriptions of the tracks.

**Comment by scottcain** [comment]
I modified the non-elegans species have the max read (where color saturates) at 1000, that combined with the 20 common/rare split seems to work well.  I also added descriptions to the "about this track".

**Comment by epaule** [comment]
B.malayi is all black ... but else looks pretty

**Comment by scottcain** [comment]
@epaule You might have a caching issue, particularly if you were using B. malayi's JBrowse yesterday while I was working on these tracks (even without loading these specific tracks).  Can you please try clearing your cache and seeing if it gets rectified?

**Comment by gw3** [comment]
Thanks @scottcain 
It looks good! :+1:

**Comment by chris-grove** [comment]
OK so this discussion has circled back to possibly displaying Denis' plots as static images on gene pages. Here's a fresh DropBox link to the supplemental PDF with the images:

https://www.dropbox.com/s/ipeuxzvjwn0yaxr/Supplemental_Fig_S2.pdf?dl=0

@tharris @sibyl229 
Would it be possible to implement a display of each image on its respective gene page? One thought was to add it to the Expression widget, but I'm open to other suggestions.
